### ISSUE TITLE
download all resources contained in one collection 

### ISSUE THREAD

**Issue by zhiyuli** [hydroshare/hydroshare#1304]
@dtarb @danames @horsburgh @alvacouch 
The requirement for downloading all resources contained in a collection was raised during the 1st phase development of collection res type. At that time it had lower priority, so several related issues that may affect the implementation have not been discussed in details (see below). About two weeks ago @dtarb mentioned CUAHSI brought up the requirement for downloading all RefTS resources contained in one collection. If this functionality is still needed, the following issues should be finalized before the implementation can proceed.

1) downloading all res bags in a collection may take a long time. in case of RefTS it may be even worse as it depends on both HS and HydroServer. Make it a asynchronous request? How long should the "big collection bag" be kept around on server? What does the "big collection bag" look like (folder structure)?

2) Since collection holds collections is allowed, do we need to download sub-collections?
"duplicated contained resources": The same Res A exists in multiple sub-collections
"duplicated sub-collections": The same sub-collection exists on multiple levels
"loops": A holds B,  B holds C, C holds A

3) permission issue: user with permission over the "root collection" may not have permission over some sub-collections or contained resources. 

4) other related issues (need your input)...

**Comment by alvacouch** [comment]
@dtarb @danames @horsburgh @zhiyuli 
The key to downloading collections is to have a consistent directory structure. This limits what one can do. Also, it is fairly obvious that loop detection is important. Might I suggest: 
```
resource/
              sciencemetadata.xml
              1st-resource/
                              1st-resource-file.txt
                                    ...
              2nd-resource/ 
                              2nd-resource.file.txt
                                     ...
``` 
This is recursive, in the sense that resource that are collections go into subdirectories. 

With this convention in hand, the question is what users want.

**Comment by alvacouch** [comment]
@dtarb @zhiyuli @horsburgh @danames 
Also, since the titles of resources aren't unique, we need to add some disambiguator to the directories of resources. Might I suggest a simple prepended count (e.g., 0001-{existing title}, 0002-{existing title}, .....
to make sure that titles are unique?

**Comment by horsburgh** [comment]
@alvacouch - Resource titles could be quite long, not guaranteed to be unique, and likely won't be suitable for directory names.  One could slugify them to shorten and prepend a number as you suggest. Another option is to just use the resource identifiers as the directory names.

**Comment by zhiyuli** [comment]
@horsburgh @alvacouch @danames @dtarb 
collection is a res in hs, how about we continue using existing bag structure for collection.

currently if user downloads a collection zip bag, he got something like this:
5b9a87d8aae74198bf244ee394d45c17  <-- root folder
 .\bagit.xt
 .\manifest-md5.txt
 .\tagmanifest-md5.txt
 .\data\
 .\data\resourcemap.xml
 .\data\resourcemetadata.xml

We could add a "contents" folder under the "data" folder as other res types do. And put all contained resource zip bags (do not unzip them) under the "content" folder:
 .\data\contents\
 .\data\contents\res_1_id.zip
 .\data\contents\res_2_id.zip

if the contained res is a collection, we unzip its zip bag, add a "contents" folder, and put its contained res zip bags into its "contents" folder.

Finally, we zip up the root folder (root collection) and delivery it to user

**Comment by horsburgh** [comment]
@zhiyuli - I really like the suggestion of just sticking with the BagIt packaging scheme for our resource data model and nesting "Collected" resources within the "Contents" folder of the collection resource.  It makes a lot of sense.  Some of these could get quite large...

**Comment by zhiyuli** [comment]
@horsburgh a second thought: Will users like to download a very big collection zip bag, unzip it, go to a deep folder, unzip several other zip bags, then go to deeper folders to see the res files....

**Comment by danames** [comment]
@zhiyuli let's try for simplicity. When a user downloads a collection they
get a master zip file. In it are contained all of the bags of the collected
resources. Very simple..

**Comment by horsburgh** [comment]
@danames - the BagIt format and the capability of the resource data model to express relationships among objects in a resource (e.g., within the collection) would be lost.  One might eventually conceive of a structured collection containing a specific set of resources related in specific ways.

A collection is a resource.  All of our resources conform to the resource data model.  I get the simplicity argument, but a BagIt bag is already simply a zip file with a little bit of internal structure.  A collection resource should already have a science metadata document (which would also need to be included in any download of the collection) and a resource map (which would also need to be included).  So - it doesn't make sense to me not to package it in the same way as all other resources.

**Comment by zhiyuli** [comment]
@horsburgh @alvacouch @danames @dtarb 
In addition to discussing "collection bag structure", 
I think a bigger issue is what the workflow of downloading a collection looks like?
Current policy is that user may download a res bag from landing page which is less than 1GB. For res larger than 1GB, user should download it from irods. A collection is much easier to become larger than 1GB compared to regular res. Can we make a simple rule: any collection bag should be downloaded from user irods space? (is there any technical barrier doing this? @hyi @rayi113 @pkdash)

**Comment by horsburgh** [comment]
@zhiyuli - if the resulting collection bag is going to be really large, we could move to a staging/queuing system where the resource is staged somewhere (perhaps in iRODS) and then an email is sent to the user with a link or instructions for how to download.  Ultimately, if the user asks to download a very large collection resource (say multiple GB) they need to be notified that it is very large and they need to be prepared to potentially wait a little bit until it has been staged for them to download.

**Comment by horsburgh** [comment]
@zhiyuli - to answer your question about zips within zips - consistent with the BagIt packaging specification, I think only the entire collection should be zipped for transport, but the contents of the resource should not be zipped.

**Comment by zhiyuli** [comment]
just had a chat with @mjstealey. A new issue #1311 was created for researching how to download big file/res. Although issue #1311 is not specific to collection, its research result would highly contribute to resolving this issue.

**Comment by horsburgh** [comment]
@zhiyuli - for an initial implementation, I don't think I would depart from the existing resource data model specification. A collection is a logical group of resources. I wouldn't name any of the resource folders within the collection differently - I would name each of them using their identifier (regardless of whether it's a sub-collection resource, a time series resource, or any other type of resource).

**Comment by zhiyuli** [comment]
@hyi 
I will continue working on the "download collection" functionality. This feature is related to "user irods zone" and "async downloading". About 3 weeks ago we had a preliminary plan in a weekly call. Since you were absent from that call, before we start implementing this, I would like have some suggestions from you to make sure the plan is doable and will not conflict with your planned changes to those two features.

The basic idea is 
1) If the collection is less than 1GB, user should be able to download it from the landing page just like downloading other regular resource bags. I am assuming this operation is synchronized.

2) If the collection is larger than 1GB, user should set up his own "user irods space" account first (if he doesn't have one), and system should check his irods space has enough space to store the collection, then the collection bag will be delivered to his irods space. This process should be asynchronized. An email will be sent to the user when the collection bag is ready in his irods space.

Thanks

**Comment by hyi** [comment]
@zhiyuli Thanks for the update. I will work on making https://github.com/hydroshare/django_irods/blob/dc76b07605a5287af8640d1c603bd8ff63b2992f/views.py#L86 call asynchronously so that it will not hold up the main web thread which may result in timeout and shutdown web server for large files that takes a long time to create the zip bag file. This requires to write front-end JavaScript code to notify users while the asyncronous task is running and update the status as the task is running. This work will only affect single file or single resource bag download in download() function here: https://github.com/hydroshare/django_irods/blob/dc76b07605a5287af8640d1c603bd8ff63b2992f/views.py#L24, which is used by collection download as well. I can add the functionality to copy the file or resource bag to user irods space as you described in No. 2 above, which will apply to all file or resource download, not only to collection resource type. Is that sufficient or there is more to it to collection resource? Specifically, does collection resource bag include all contained resource bags? Currently, if the downloaded file or bag is larger than 1GB, we inform the user to download the large file via iRODS clients. Perhaps we can add more informative message asking them to click on "Replicate resource bag to irods user zone" button to replicate resource bag first. Do we really need to automate this process as an asychronous task as you described above for all resources? I am thinking it may be better to inform the user and ask user to take actions (just one click of button to replicate resource bag to irods user zone) rather than doing it automatically since the user may not want to download the large bag after learning the bag can not be downloaded directly due to its large size. I am looping in @dtarb and @danames to get their inputs on this as well.

**Comment by dtarb** [comment]
I am not sure exactly what the question for me is.  I think that if the total collection is less than the size that can be downloaded (1 GB) the user should just click download and get the zipped collection.  If the total collection is greater than the size that can be downloaded the direct download button should be disabled with a hover over message indicating too big to download suggesting they replicate to iRODS and download using an iRODS client.

**Comment by hyi** [comment]
@dtarb OK. That answered my question and is what is currently being implemented with slight variation for bag download due to on-demand bagging. Since the bag may not exist until being requested to download, we don't know the size of the bag until the bag is created, so we cannot disable download button in advance until the user clicks on it to request bag creation. So if after the bag is created and its size is larger than 1GB after the user clicks on download button, the user will get a response notification indicating that file larger than 1GB cannot be downloaded directly via HTTP and they can download the large file via iRODS clients.

**Comment by zhiyuli** [comment]
@hyi Currently the collection resource bag stored on hydroshare irods (main zone) is just a very small (almost empty) bag as it only has some xml files (metadata of this collection res) in it. When user clicks the "download button" on a collection resource landing page, this small/empty bag will be zipped up and delivered to user just like downloading other res types. We don't want to change the function of this button.

The preliminary plan was to create a new button ("download collection bag with all contained resource bag in it") on collection res landing page, and this new button should do the job what is described in (1) and (2) in my previous post, which means this new button should assembly the big collection bag on the fly (put all contained resource bags into this collection bag), and either (1)return the zipped big bag to user directly from landing page or (2)copy it over to user irod space and inform user to download it from there.

**Comment by zhiyuli** [comment]
@hyi 
If I understood correctly, your changes will make the "download buttons" on all resource landing pages work asynchronously?

**Comment by hyi** [comment]
@zhiyuli OK. Got it. It is a separate function than the generic download function that apply to all bags then. Currently I am only planning to make create_bag_by_irods() run asynchronously so that it will not block the main web thread especially for big files. The large files (larger than 1GB) will not be downloaded via http but return response that informs the user to replicate large files to their user irods space and download the large file via iRODS clients instead.

**Comment by zhiyuli** [comment]
@hyi @rayi113 @pkdash @mjstealey 

I added some background to this issue.

On the hydroshare Zone, the bag of a collection resource has metadata only (no contained resource bag is stored in it), so it is kind of a "empty bag". We won't move/change this "empty bag".

When user wants to download a collection bag with all contained resources in it (we call this collection bag "big bag"), we need to assemble the "big bag" on the fly, which includes the following steps:
1) make a clone of the original "empty bag"
2) put all contained resource bags into this replicated "empty bag", so it becomes a "big bag". The original "empty bag" does not change.
3) zip this "big bag"

The question is:
Where should the 3 steps take place (where should the resulting "big bag" stored initially)? on hydroshare zone or irods user zone? 

Option A: Store on hydroshare zone:
1) If the "big bag" is bigger than 1GB, we need to move it to user zone;
2) We always have a "big bag" on hydroshare zone which may be reusable if other users want ot download the same collection "big bag".

Option B: Store on user irods zone:
1) We dont need to move it even if it is bigger than 1GB beacaue it is created/assembled here initially
2) user could change/delete this "big bag". We need to re-assembly this "big bag" if other user wants to download the same collection.

In either option, the original "empty bag" on hydroshare zone is protected.

**Comment by dtarb** [comment]
I think that option B above is better.  The main reason is that the contents of the "big bag" are user dependent, because the user downloading the collection may not have rights on all the resources in the collection, with resources without rights to hold a README.txt file as @alvacouch suggested above.  Thus I do not see a viable option for big bag reuse.

I suggest the button be labeled "Copy resources in collection to iRODS".  I suggest the download button be labeled "Download collection listing as zipped Bagit Archive" to make it clear that the user is just getting a listing.  I suggest that the collection listing download include, in addition to the xml metadata files, a plain text file in the contents folder named "CollectionResourceList.txt" that contains in CSV the title, GUID, Owners (if there are multiple list all separate them by comma's and put the whole list in quotes to be CVS consistent) of each resource in the collection.

The user would then have to use an iRODS client to retrieve the contents.  I think we have to accept this, even for small collections.  This does raise the question though, of whether the big bag in iRODS should be zipped.  It may be easier to just leave the resources as folders.  Answering this depends on iRODS client capability for zipping/unzipping.  I would defer to those more knowledgeable in iRODS for that.

@zhiyuli I would encourage you to not hesitate to ask @horsburgh any questions related to the resources data model to increase the chances of having your implementation consistent with other resources.

**Comment by zhiyuli** [comment]
@dtarb Good point. The contents of the "big bag" are indeed user dependent.

**Comment by horsburgh** [comment]
@zhiyuli - I had a a related discussion with @dtarb yesterday.  One way to simplify this is to only allow collections of public and/or published resources. That removes the user dependency of the "big bag" because all of the contents would be available to all users regardless because they are public or published.  

If users want to group private resources, they should probably be using tags and/or folders (once we have enabled them on the user interface) - not collections.

The only complexity at that point would be what to do if a public resource in a collection transitions to private (say, because a user deletes the resource files, which automatically makes the resource private).

**Comment by dtarb** [comment]
Added low priority in a pass over issues to prioritize because I think that bulk downloads should be implemented not specific to collection resource but to any set of selected resources on discover or my resources page.  Then selecting the resource in a collection could be a special case of that functionality.  However I will await possible changes as collections are discussed further.

**Comment by eclark-cuahsi** [comment]
closing per backlog review meeting on 6/9/22. stale.
